trio/_subprocess.py

```diff
@@ -542,7 +614,20 @@ async def read_output(stream, chunks):
                 nursery.start_soon(read_output, proc.stdout, stdout_chunks)
             if proc.stderr is not None:
                 nursery.start_soon(read_output, proc.stderr, stderr_chunks)
-            await proc.wait()
+            try:
+                await proc.wait()
+            except trio.Cancelled:
+                with trio.CancelScope(shield=True):
+                    killer_cscope = trio.CancelScope(shield=True)
+
+                    async def killer():
+                        with killer_cscope:
+                            await deliver_cancel(proc)
```

Comment:
What should happen if deliver_cancel raises an exception? In the current implementation, the shielded cancel scope will prevent it from propagating until the process exits, which might take a while if the crash occurred before signaling the process in any way. Maybe on exception from a user-specified deliver_cancel we should call the default deliver_cancel to kill the process? Or just kill() since we don't care much about polish at that point, only about letting the user know what they need to fix.

Reply:
It isn't immediately obvious to me what the right answer is here, so splitting it off into #1532